### PR TITLE
Fixing StableHLO->linalg assumption on element attrs type.

### LINKED ISSUE #14415: Elided elements attributes cause errors even with `--iree-util-zero-fill-elided-attrs`
I rely on `--mlir-elide-elementsattrs-if-larger=10` as a first step in reducing large testcases.

Let me know if I'm missing something, but I thought that the resulting bytecode modules should be accepted by `iree-compile --iree-util-zero-fill-elided-attrs`. They're usually not at the moment; this issue seems to happen so consistently over the past few cases I've deal with, both f32 and i8 models, very different from each other.

Just as one instance, taking the testcase from https://github.com/openxla/iree/issues/14414 , doing

```
tools/iree-opt --mlir-elide-elementsattrs-if-larger=10 /tmp/stablehlo.mlirbc > /tmp/stablehlo.elided.mlir
```

then running the repro `iree-compile` command line with `--iree-util-zero-fill-elided-attrs` added:

```
$ tools/iree-compile --output-format=vm-bytecode     --mlir-print-op-on-diagnostic=false     --iree-hal-target-backends=llvm-cpu    --iree-input-type=stablehlo     --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu     --iree-llvmcpu-target-cpu=cascadelake    --iree-flow-enable-data-tiling     --iree-llvmcpu-enable-microkernels  --iree-util-zero-fill-elided-attrs   /tmp/stablehlo.elided.mlir -o /tmp/module.vmfb
```

Fails like this:

```
iree-compile: iree/third_party/llvm-project/llvm/include/llvm/Support/Casting.h:566: decltype(auto) llvm::cast(const From &) [To = mlir::DenseElementsAttr, From = mlir::ElementsAttr]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```

My workaround is to rewrite elided attribute as `dense<0>` or `dense<0.0>` depending on whether the element type is floating point. For an all-float model like the one above, I do:

```
tools/iree-opt --mlir-elide-elementsattrs-if-larger=10 /tmp/stablehlo.mlirbc | sed 's/dense_resource<__elided__>/dense<0.0>/g' > /tmp/stablehlo.elided.mlir
```

### PR DESCRIPTION
**Pull request by benvanik**
This allows elided elements attrs (which are not DenseElementAttrs) to pass through.

Also added a test for iree-util-zero-fill-elided-attrs to verify it works on its own.

Fixes #14415.

### CODE REVIEW

**Review by bjacob — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/test/constant_encoding.mlir [hunk 1]
```diff
@@ -121,6 +122,32 @@ vm.module @constants {
...
+  // during iteration.
+  //      CHECK: "embedded_data": [
+  // CHECK-NEXT:   0,
+  // CHECK-NEXT:   0,
+  // CHECK-NEXT:   0
+  // CHECK-NEXT: ]
+  vm.rodata private @elided_i2 dense_resource<__elided__> : tensor<9xi2>
+
+  //      CHECK: "embedded_data": [
+  // CHECK-NEXT:   0,
```
Comment (kuhar) [inline compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/test/constant_encoding.mlir:139]: I wonder if we can use `CHECK-COUNT-k: 0,` here

File: compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp [hunk 2]
```diff
@@ -1507,15 +1507,18 @@ class ConstConverterTensor
...
-    auto type =
+    auto replacementType =
         llvm::cast<ShapedType>(typeConverter->convertType(constOp.getType()));
-    if (type != constOp.getType()) {
-      // Signedness conversion.
-      valueAttr = valueAttr.mapValues(type.getElementType(),
-                                      [](const APInt &i) { return i; });
+    ElementsAttr replacementAttr = constOp.getValue();
+    if (replacementType != constOp.getType()) {
+      if (auto denseAttr = llvm::cast<DenseElementsAttr>(replacementAttr)) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp]: This needs `dyn_cast`. Also nit: no need for `llvm::` -- AFAIK it was added to simplify automatic upgrade to free-function-style casts.

File: compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/test/constant_encoding.mlir (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/test/constant_encoding.mlir:139]: tricky due to JSON - the trailing element has no comma so it'd have to be a CHECK-COUNT-(k-1) + CHECK-NEXT ;(
(I'd had the commas omitted for awhile but in other cases like CHECK: 1 that'd be satisfied by 1 or 10 or 100)

File: compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp (hunk 2 above)
Comment (benvanik) [reply compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp]: yeah that llvm:: bothers me, didn't know if this file preferred it but if not I'll strip here

**Review by kuhar — APPROVED** [review]
